bip-xxxx.mediawiki

```diff
@@ -0,0 +1,283 @@
+     |       +------------------+       |                      |
+     |       | Payjoin Proposal |       |                      |
+     +-------+      PSBT        +------>+                      |
+     |       +------------------+       |                      |
+     |                                  |   +--------------+   |
+     |                                  |---+ Payjoin      |   |
+     |                                  |   | transaction  +-->+
+     |                                  |   +--------------+   |
+     +                                  +                      +
+</pre>
+The original PSBT is sent in the HTTP POST request body, base64 serialized, with <code>text/plain</code> in the <code>Content-Type</code> HTTP header and <code>Content-Length</code> set correctly.
+The payjoin proposal PSBT is sent in the HTTP response body, base64 serialized with HTTP code 200.
+
+===Receiver's well known errors===
+
+If for some reason the receiver is unable to create a payjoin proposal, it will reply with a HTTP code different than 200.
+The receiver is not constrained to specific set of errors, some are specified in this proposal.
+
+The errors have the following format:
+<pre>
+{
+    "errorCode": "leaking-data",
+    "message": "Key path information or GlobalXPubs should not be included in the original PSBT."
+}
+</pre>
+
+The well-known error codes are:
+{| class="wikitable"
+!Error code
+!Meaning
+|-
+|leaking-data
+|Key path information or GlobalXPubs should not be included in the original PSBT.
+|-
+|psbt-not-finalized
+|The original PSBT must be finalized.
+|-
+|unavailable
+|The payjoin endpoint is not available for now.
+|-
+|out-of-utxos
+|The receiver does not have any UTXO to contribute in a payjoin proposal.
+|-
+|not-enough-money
+|The receiver added some inputs but could not bump the fee of the payjoin proposal.
+|-
+|insane-psbt
+|Some consistency check on the PSBT failed.
+|-
+|version-unsupported
+|This version of payjoin is not supported.
+|-
+|need-utxo-information
+|The witness UTXO or non witness UTXO is missing
+|}
+
+The receiver is allowed to return implementation specific errors which may assist the sender to diagnose any issue.
+
+However, it is important that error codes that are not well-known and that the message do not appear on the sender's software user interface.
+Such error code or message could be used maliciously to fish a non technical user.
+Instead those error or message can only appear in debug logs.
+
+It is advise to hard code the description of the error code into the sender's software.
+
+===Sender's payjoin proposal checklist===
+
+The sender should check the payjoin proposal before signing it to prevent a malicious receiver from stealing money.
+
+* Check that all the spent outpoints in the original PSBT still exists in the coinjoin PSBT.
+* Check that all the spent outpoints in the original PSBT does not have any partial signature.
+* Check that the receiver added at least one input.
```

Comment:
Just a random thought, not sure if it's baseless.

What if we remove this requirement and just allow the payjoin endpoint to redirect the outputs of the proposed transaction? 

It would allow settling payments scheduled by the receiver by using the sender's funds and not require any UTXOs at hand, and still poison heuristics even more. A 1 input tx could even be used to pay the receiver, and the receiver uses the same tx to pay scheduled payments. It is technically not a payjoin, but it is one in the world of meta as you've redirected funds from User A to Merchant A to MerchantSupplier A and MerchantSupplier B, with only 1 transaction.

Reply:
![image](https://user-images.githubusercontent.com/1818366/81499677-79f0da00-92cd-11ea-8598-ef6c14bbb9df.png)

Reply:
This is a great idea.

Reply:
I see this too late. The idea sounds good but the sender should reject it because it gains no privacy.

Reply:
@lontivero it does. The receiver can make the change harder to detect.